### PR TITLE
ipc4: fix volume issues on windows platform

### PR DESCRIPTION
**Pull request by RanderWang**
Three kinds of issues found on windows
(1) FW error
(2) FW panic
(3) volume can't be changed

(1)(2) are caused by none fade config (3) we need set|get large config

One issue on Linux:
gain is not supported

### CODE REVIEW

File: src/audio/volume/volume.c [hunk 1]
```diff
@@ -929,6 +963,91 @@ static int volume_cmd(struct comp_dev *dev, int cmd, void *data,
...
+					true);
+			if (cd->volume[cdata->channel_id] != cd->tvolume[cdata->channel_id])
+				cd->ramp_finished = false;
+		}
+
+		prepare_ramp(dev, cd);
+		break;
+
+	default:
+		return -EINVAL;
```
Comment (lgirdwood) [inline src/audio/volume/volume.c:1019]: If the param is invalid it's best to report this in the trace.

File: src/audio/volume/volume.c [hunk 2]
```diff
@@ -929,6 +963,91 @@ static int volume_cmd(struct comp_dev *dev, int cmd, void *data,
...
+		for (i = 0; i < cd->channels; i++) {
+			uint32_t volume = cd->peak_regs.target_volume_[i];
+
+			cdata[i].channel_id = i;
+			cdata[i].target_volume = convert_ipc3_vol_to_ipc4(volume);
+		}
+		*data_offset = sizeof(*cdata) * cd->channels;
+		break;
+	default:
+		return -EINVAL;
```
Comment (lgirdwood) [inline src/audio/volume/volume.c:1050]: ditto

**Review by singalsu — CHANGES_REQUESTED** [review]

**Comment by singalsu** [comment]
@RanderWang Kudos for remembering to round the math operations!

File: src/audio/volume/volume.c [hunk 3]
```diff
@@ -428,11 +432,101 @@ static int set_volume_ipc4(struct vol_data *cd, uint32_t const channel,
...
 #endif
 	return 0;
 }
 
+/* In IPC4 driver sends volume in Q1.31 format. It is converted
+ * into Q1.23 format to be processed by firmware.
+ */
+static inline uint32_t convert_ipc4_vol_to_ipc3(struct comp_dev *dev, uint32_t volume)
+{
+	volume = Q_SHIFT_RND(volume, 31, 23);
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/audio/volume/volume.c:447]: sure, thanks

File: src/audio/volume/volume.c [hunk 4]
```diff
@@ -428,11 +432,101 @@ static int set_volume_ipc4(struct vol_data *cd, uint32_t const channel,
...
+
+	return volume;
+}
+
+static inline uint32_t convert_ipc3_vol_to_ipc4(uint32_t volume)
+{
+	/* In IPC4 volume is converted into Q1.23 format to be processed by firmware.
+	 * Now convert it back to Q1.31
+	 */
+	volume = Q_SHIFT_LEFT(volume, 23, 31);
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/audio/volume/volume.c:476]: thanks

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@RanderWang can you check CI, lots of failures.

**Comment by lgirdwood** [comment]
@singalsu good now ?

File: src/include/ipc/topology.h [hunk 5]
```diff
@@ -152,6 +152,7 @@ enum sof_volume_ramp {
 	SOF_VOLUME_LINEAR_ZC,
 	SOF_VOLUME_LOG_ZC,
 	SOF_VOLUME_WINDOWS_FADE,
+	SOF_VOLUEM_WINDOWS_NO_FADE,
```
Comment (singalsu) [inline src/include/ipc/topology.h:155]: Topo in enum,  must fix. Otherwise good!

**Review by singalsu — CHANGES_REQUESTED** [review]

File: src/include/ipc/topology.h (hunk 5 above)
Comment (RanderWang) [reply src/include/ipc/topology.h:155]: :)

**Review by marcinszkudlinski — APPROVED** [review]

File: src/audio/volume/volume.c [hunk 6]
```diff
@@ -1070,7 +1070,7 @@ static int volume_prepare(struct comp_dev *dev)
 		goto err;
 	}
 
-	if (!cd->ramp_func) {
+	if (cd->initial_ramp && !cd->ramp_func) {
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/audio/volume/volume.c:1215]: yes,  windows supports none ramp config

File: src/audio/volume/volume.c [hunk 7]
```diff
@@ -433,6 +433,28 @@ static int set_volume_ipc4(struct vol_data *cd, uint32_t const channel,
 	return 0;
 }
 
+/* In IPC4 driver sends volume in Q1.31 format. It is converted
+ * into Q1.23 format to be processed by firmware.
+ */
+static inline uint32_t convert_ipc4_vol_to_ipc3(struct comp_dev *dev, uint32_t volume)
```
[the remark this answers is not in the document]
Comment (RanderWang) [reply src/audio/volume/volume.c:445]: sure, thanks

**Review by singalsu — APPROVED** [review]

**Comment by RanderWang** [comment]
updated, thanks

**Pull request merged**